src/mpid/ch4/netmod/ofi/ofi_am_impl.h

```diff
@@ -505,7 +505,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_am_isend_eager(int rank, MPIR_Comm * c
         MPIR_GPU_query_pointer_attr(buf, &attr);
         if (attr.type == MPL_GPU_POINTER_DEV) {
             MPIDI_OFI_register_am_bufs();
-            if (!MPIDI_OFI_ENABLE_HMEM) {
+            if (!MPIDI_OFI_ENABLE_HMEM || !MPIR_CVAR_CH4_OFI_ENABLE_HMEM) {
```

Comment:
I don't think its necessary to check both these conditions, is it? My understanding is that `MPIDI_OFI_ENABLE_HMEM` is updated during init, which takes `MPIR_CVAR_CH4_OFI_ENABLE_HMEM` into account. For example, we do not check both `MPIDI_OFI_ENABLE_TAGGED` and `MPIR_CVAR_CH4_OFI_ENABLE_TAGGED` in the send path.

Reply:
From my understanding, `MPIDI_OFI_ENABLE_HMEM` is statically defined as to whether MPICH should expect HMEM support for the given provider. For example, this is set to `MPIDI_OFI_ON` only for CXI provider currently. However `MPIR_CVAR_CH4_OFI_ENABLE_HMEM` is user-tunable knob on whether the user wants to enable GPU RDMA in the provider (if it is supported).

So checking for both ensures it is supported in the provider and the user wants it turned on. In this particular case, if HMEM is supported and the user enables `MPIR_CVAR_CH4_OFI_ENABLE_HMEM`, then GPU RDMA path is taken. Otherwise, if either of these are not true, then D2H path should be taken. If we just checked for `MPIDI_OFI_ENABLE_HMEM`, then the CVAR knob would not be considered and GPU RDMA would always be taken for providers that support it.

Reply:
`MPIR_CVAR_CH4_OFI_ENABLE_HMEM` is checked during init time and is reflected in `MPIDI_OFI_ENABLE_HMEM`, at least it should be.

Reply:
After discussing with @zhenggb72, it seems that in optimized builds (i.e. selecting a provider via `--with-device=cxi`) we lose the ability to turn off GPU RDMA capability because runtime capability selection is disabled. We would like to preserve the ability to toggle off GPU RDMA for optimized builds, which is the reason for checking both

Reply:
>  it seems that in optimized builds (i.e. selecting a provider via --with-device=cxi) we lose the ability to turn off GPU RDMA capability because runtime capability selection is disabled. 

Then we should fix that. I'll look into it.

Reply:
In the meantime, should I keep the additional checks for `MPIR_CVAR_CH4_OFI_ENABLE_HMEM` or remove them?

Reply:
Let's see if this https://github.com/pmodels/mpich/pull/6601 will work

Reply:
Thanks @hzhou! These changes look good to me. I'll start removing the CVAR check in favor of these changes

Reply:
Rebased and removed the redundant CVAR checks